torch/csrc/tensor/python_tensor.cpp

```diff
@@ -0,0 +1,261 @@
+#include "python_tensor.h"
+
+#include <structmember.h>
+#include <mutex>
+#include <pybind11/pybind11.h>
+#include <sstream>
+
+#include "torch/csrc/assertions.h"
+#include "torch/csrc/Exceptions.h"
+#include "torch/csrc/autograd/variable.h"
+#include "torch/csrc/autograd/python_variable.h"
+#include "torch/csrc/autograd/generated/VariableType.h"
+#include "torch/csrc/utils/python_strings.h"
+#include "torch/csrc/utils/tensor_new.h"
+#include "torch/csrc/utils/tensor_types.h"
+
+namespace torch { namespace tensor {
+
+using namespace at;
+using namespace torch::autograd;
+
+struct PyTensorType {
+  PyTypeObject py_type;
+  at::Type* aten_type;
+  bool is_cuda;
+  bool is_sparse;
+  bool is_default;
+  char name[64];
+};
+
+static_assert(std::is_standard_layout<PyTensorType>::value, "PyTensorType must be standard layout");
+
+static PyTensorType* default_tensor_type;
+static std::once_flag init_cuda_flag;
+
+static void py_bind_tensor_types(const std::vector<PyTensorType>& tensor_types);
+
+static PyObject* Tensor_new(PyTypeObject *type, PyObject *args, PyObject *kwargs) {
+  HANDLE_TH_ERRORS
+  auto& tensor_type = *((PyTensorType*)type);
+  if (!tensor_type.aten_type) {
+    throw TypeError("type %s not available", tensor_type.name);
```

Comment:
probably good to special case an error message about cuda;  There's an example in python_arg_parser:
    if (dtype->is_cuda) {
      oss << "  Torch not compiled with CUDA enabled." << std::endl;
    }